component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/jwt/JWTBearerGrantHandler.java

```diff
@@ -758,12 +882,100 @@ protected X509Certificate resolveSignerCertificate(JWSHeader header,
+                    break;
+                }
+            }
+        }
+        if (StringUtils.isEmpty(assertion)) {
+            if (log.isDebugEnabled()) {
+                log.debug("The assertion is empty.");
+            }
+            return null;
+        }
+
+        try {
+            EncryptedJWT encryptedJWT = EncryptedJWT.parse(assertion);
+            return encryptedJWT;
+        } catch (ParseException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Error while parsing the assertion. The assertion is not encrypted.");
+            }
+            return null;
+        }
+    }
+
+    private static RSAPrivateKey getPrivateKey(String tenantDomain) throws IdentityOAuth2Exception {
+
+        Key privateKey;
+        int tenantId = OAuth2Util.getTenantId(tenantDomain);
+        if (!(privateKeys.containsKey(tenantId))) {
+
+            try {
+                IdentityTenantUtil.initializeRegistry(tenantId, tenantDomain);
+            } catch (IdentityException e) {
+                throw new IdentityOAuth2Exception("Error occurred while loading registry for tenant " + tenantDomain,
+                        e);
+            }
+            // get tenant's key store manager
+            KeyStoreManager tenantKSM = KeyStoreManager.getInstance(tenantId);
+
+            if (!MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                // derive key store name
+                String ksName = tenantDomain.trim().replace(".", "-");
+                String jksName = ksName + ".jks";
+                // obtain private key
+                privateKey = tenantKSM.getPrivateKey(jksName, tenantDomain);
+
+            } else {
+                try {
+                    privateKey = tenantKSM.getDefaultPrivateKey();
+                } catch (Exception e) {
+                    //Intentionally catch Exception as an Exception is thrown from the above layer.
+                    throw new IdentityOAuth2Exception("Error while obtaining private key for super tenant", e);
+                }
+            }
+            //privateKey will not be null always
+            privateKeys.put(tenantId, privateKey);
+        } else {
+            //privateKey will not be null because containsKey() true says given key is exist and ConcurrentHashMap
+            // does not allow to store null values
+            privateKey = privateKeys.get(tenantId);
+        }
+        return (RSAPrivateKey) privateKey;
+    }
+
+    private boolean isEncryptedJWTSigned(String payload) {
+
+        if (StringUtils.isNotEmpty(payload)) {
+            String[] parts = payload.split(".");
+            if (parts.length == 3) {
+                if (StringUtils.isNotEmpty(parts[2].toString())) {
+                    return true;
+                }
+                return false;
+            }
+            return false;
+        } else {
+            return false;
+        }
+    }
+}
```

Comment:
How about writing the above method as below. It will be more readable IMO. :)

```
private boolean isEncryptedJWTSigned(String payload) {

    if (StringUtils.isNotEmpty(payload)) {
        String[] parts = payload.split(".");
        if (parts.length == 3 && StringUtils.isNotEmpty(parts[2].toString())) {
            return true;
        }
    }
    return false;
}
```